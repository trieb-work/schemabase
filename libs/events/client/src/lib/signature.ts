import { createHmac } from "crypto";
import { env } from "@chronark/env";
import { ILogger } from "@eci/util/logger";
import { SignatureError } from "./errors";

export interface ISigner {
  sign(data: unknown): string;
  verify(data: unknown, signature: string): void;
}

/**
 * Sign and verify json objects
 */
export class Signer implements Signer {
  /**
   * Secret key used to sign the data.
   */
  private readonly signingKey: string;

  constructor(logger: ILogger) {
    if (env.get("NODE_ENV") === "production") {
      this.signingKey = env.require("SIGNING_KEY");
    } else {
      logger.warn(
        "Using generated signing key. In production you need to provide the `SIGNING_KEY` env variable",
      );
      this.signingKey = "super-secret-key";
    }
  }

  private serialize(data: unknown): string {
    return JSON.stringify(data);
  }
  /**
   * Create a signature for the given data.
   * @param data - object to be signed
   * @returns
   */
  public sign(data: unknown): string {
    return createHmac("sha256", this.signingKey)
      .update(this.serialize(data))
      .digest("hex");
  }

  /**
   * Verify the data was signed by the given signature.
   */
  public verify(data: unknown, expectedSignature: string): void {
    const signature = this.sign(data);
    if (signature !== expectedSignature) {
      throw new SignatureError();
    }
  }
}
