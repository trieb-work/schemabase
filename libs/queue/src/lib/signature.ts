import { createHmac, randomUUID } from "crypto";
import { env } from "@chronark/env";
import { Logger } from "@eci/util/logger";

/**
 * Sign and verify json objects
 */
export class Signer {
  /**
   * Secret key used to sign the data.
   */
  private readonly signingKey: string;

  constructor(logger: Logger) {
    if (env.get("NODE_ENV") === "production") {
      this.signingKey = env.require("SIGNING_KEY");
    } else {
      logger.warn(
        "Using generated signing key. In production you need to provide the `SIGNING_KEY` env variable",
      );
      this.signingKey = randomUUID();
    }
  }

  private serialize(data: unknown): string {
    return JSON.stringify(data);
  }
  /**
   * Create a signature for the given data.
   * @param data - object to be signed
   * @returns
   */
  public sign(data: unknown): string {
    return createHmac("sha512", this.signingKey)
      .update(this.serialize(data))
      .digest("hex");
  }

  /**
   * Verify the data was signed by the given signature.
   */
  public verify(data: unknown, expectedSignature: string): boolean {
    const signature = this.sign(data);
    return signature === expectedSignature;
  }
}
