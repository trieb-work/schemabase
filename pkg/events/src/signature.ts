import { createHmac } from "crypto";
import { SignatureError } from "./errors";

export interface ISigner {
    sign: (data: unknown) => string;
    verify: (data: unknown, signature: string) => void;
}

export interface SignerConfig {
    signingKey: string;
}

/**
 * Sign and verify json objects
 */
export class Signer implements ISigner {
    /**
     * Secret key used to sign the data.
     */
    private readonly signingKey: string;

    constructor({ signingKey }: SignerConfig) {
        this.signingKey = signingKey;
    }

    /**
     * Create a signature for the given data.
     * @param data - object to be signed
     * @returns
     */
    public sign(data: unknown): string {
        return createHmac("sha256", this.signingKey)
            .update(JSON.stringify(data))
            .digest("hex");
    }

    /**
     * Verify the data was signed by the given signature.
     */
    public verify(data: unknown, expectedSignature: string): void {
        const signature = this.sign(data);
        if (signature !== expectedSignature) {
            throw new SignatureError(signature, expectedSignature);
        }
    }
}
