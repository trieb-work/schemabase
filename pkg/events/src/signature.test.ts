import { describe, expect, it } from "@jest/globals";
import { Signer } from "./signature";

const testCases: { name: string; message: unknown; signature: string }[] = [
  {
    name: "objects",
    message: {
      hello: "world",
    },
    signature:
      "2677ad3e7c090b2fa2c0fb13020d66d5420879b8316eb356a2d60fb9073bc778",
  },
  {
    name: "strings",
    message: "I am a string",
    signature:
      "61185ff480c83442fd597f087be0b9b17829f3753bf5c676ad3fec162d59543d",
  },
  {
    name: "numbers",
    message: 123,
    signature:
      "77de38e4b50e618a0ebb95db61e2f42697391659d82c064a5f81b9f48d85ccd5",
  },
  {
    name: "booleans",
    message: false,
    signature:
      "6a20f9fc3e1489885684d050218266e98e8a932c8ebe27cfbf97ca05f992c5dd",
  },
  {
    name: "arrays",
    message: ["no", 1, true],
    signature:
      "41f423075c35fa8026b88912a98ab56950efd6279527c38d9aba82b18e8dc89a",
  },
];
const signer = new Signer({ signingKey: "secret" });

describe("sign()", () => {
  for (const tc of testCases) {
    it(`should create the correct signature for ${tc.name}`, () => {
      const signature = signer.sign(tc.message);
      expect(signature).toEqual(tc.signature);
    });
  }
});

describe("verify()", () => {
  describe("when the signature is correct", () => {
    for (const tc of testCases) {
      it(`should pass for ${tc.name}`, () => {
        expect(() => signer.verify(tc.message, tc.signature)).not.toThrow();
      });
    }
  });

  describe("when the signature does not match", () => {
    it("should throw", () => {
      expect(() =>
        signer.verify("hello", "wrongSignature"),
      ).toThrowErrorMatchingInlineSnapshot(
        '"Signature is invalid, got: 6ae6caf33c4be33bbf6c699b171bd8c2b85b56778c99a425d0f41461f08fe222, expected: wrongSignature"',
      );
    });
  });
});
