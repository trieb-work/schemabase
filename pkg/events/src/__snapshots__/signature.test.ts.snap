// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify() when the signature does not match should throw 1`] = `"Signature is invalid, got: 6ae6caf33c4be33bbf6c699b171bd8c2b85b56778c99a425d0f41461f08fe222, expected: wrongSignature"`;
